### PR TITLE
Have command line argument to set region_pg_cnt

### PR DESCRIPTION
**Pull request by nkottary**
This lets the user provide a different virtual memory size than the default.

### CODE REVIEW

File: src/jloptions.c [hunk 1]
```diff
@@ -505,6 +507,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
             else
                 jl_errorf("julia: invalid argument to --handle-signals (%s)", optarg);
             break;
+        case 'r':
+            {
+                int rpc = strtol(optarg, &endptr, 10);
+                if (errno != 0 || optarg == endptr) {
```
Comment (yuyichao) [inline src/jloptions.c]: You should never check `errno` to determine if an error happens (in general).

File: src/jloptions.c [hunk 2]
```diff
@@ -51,7 +51,8 @@ jl_options_t jl_options = { 0,    // quiet
                             NULL, // output-o
                             NULL, // output-ji
                             0, // incremental
-                            0 // image_file_specified
+                            0, // image_file_specified
+                            0  // region-pg-cnt
 };
 
 static const char usage[] = "julia [switches] -- [programfile] [args...]\n";
```
Comment (yuyichao) [inline src/jloptions.c]: Need to add to help string too.

File: src/jloptions.c [hunk 3]
```diff
@@ -505,6 +507,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
             else
                 jl_errorf("julia: invalid argument to --handle-signals (%s)", optarg);
             break;
+        case 'r':
+            {
+                int rpc = strtol(optarg, &endptr, 10);
+                if (errno != 0 || optarg == endptr) {
+                    jl_errorf("julia: -r,--region-pg-cnt=<n> must be an integer >= 1");
+                } else if (*endptr == 'm') {
+                    jl_options.region_pg_cnt = 64 * rpc;
```
Comment (yuyichao) [inline src/jloptions.c]: It's a little weird that the optional is called page count but the value is in size.

File: src/jloptions.c [hunk 4]
```diff
@@ -505,6 +507,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
             else
                 jl_errorf("julia: invalid argument to --handle-signals (%s)", optarg);
             break;
+        case 'r':
```
Comment (yuyichao) [inline src/jloptions.c]: I don't know what's exactly the use case so I'm fine with an option. Just note that this will not propagate automatically to workers or other processes you starts.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: src/jloptions.c [hunk 5]
```diff
@@ -505,6 +507,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
...
+            {
+                int rpc = strtol(optarg, &endptr, 10);
+                if (errno != 0 || optarg == endptr) {
+                    jl_errorf("julia: -r,--region-pg-cnt=<n> must be an integer >= 1");
+                } else if (*endptr == 'm') {
+                    jl_options.region_pg_cnt = 64 * rpc;
+                } else if (*endptr == 'g') {
+                    jl_options.region_pg_cnt = 65536 * rpc;
+                } else {
+                    jl_errorf("julia: -r,--region-pg-cnt=<n> Invalid unit provided. Valid units are 'm' (mega) and 'g' (giga). For example '10m'");
```
Comment (yuyichao) [inline src/jloptions.c]: `M`, and `G` should be the more standard spelling.

File: src/jloptions.c [hunk 6]
```diff
@@ -505,6 +507,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
             else
                 jl_errorf("julia: invalid argument to --handle-signals (%s)", optarg);
             break;
+        case 'r':
+            {
+                int rpc = strtol(optarg, &endptr, 10);
+                if (errno != 0 || optarg == endptr) {
+                    jl_errorf("julia: -r,--region-pg-cnt=<n> must be an integer >= 1");
+                } else if (*endptr == 'm') {
```
Comment (yuyichao) [inline src/jloptions.c]: Should check if there's garbage after this too.

File: src/jloptions.c [hunk 7]
```diff
@@ -505,6 +507,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
             else
                 jl_errorf("julia: invalid argument to --handle-signals (%s)", optarg);
             break;
+        case 'r':
+            {
+                int rpc = strtol(optarg, &endptr, 10);
+                if (errno != 0 || optarg == endptr) {
+                    jl_errorf("julia: -r,--region-pg-cnt=<n> must be an integer >= 1");
```
Comment (yuyichao) [inline src/jloptions.c]: And maybe also check the value against min-count so that the option is not silently ignored.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: src/jloptions.c [hunk 8]
```diff
@@ -66,6 +67,7 @@ static const char opts[]  =
     " -H, --home <dir>          Set location of `julia` executable\n"
     " --startup-file={yes|no}   Load ~/.juliarc.jl\n"
     " --handle-signals={yes|no} Enable or disable Julia's default signal handlers\n\n"
+    " -s, --page-size={N}       Set the GC page size\n\n"
```
Comment (yuyichao) [inline src/jloptions.c]: It's not page size, but region size.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: src/jloptions.c (hunk 1 above)
Comment (yuyichao) [inline src/jloptions.c]: Not addressed yet.

File: src/jloptions.c (hunk 4 above)
Comment (yuyichao) [inline src/jloptions.c]: Still better to figure out if this should propagate.

File: src/jloptions.c (hunk 7 above)
Comment (yuyichao) [inline src/jloptions.c]: Not addressed yet.

File: src/gc-pages.c [hunk 9]
```diff
@@ -24,6 +24,9 @@ static size_t current_pg_count = 0;
 
 void jl_gc_init_page(void)
 {
+    if (jl_options.region_pg_cnt >= MIN_REGION_PG_COUNT) {
```
Comment (vtjnash) [inline src/gc-pages.c]: If less than the minimum, it should probably bump it up equal to the minimum rather than ignoring it

File: src/jloptions.c (hunk 8 above)
Comment (vtjnash) [inline src/jloptions.c]: This probably shouldn't have a short name. It might also be nicer to group these somehow. Something like `--config.gc.region.size={N}` ?

File: src/jloptions.c [hunk 10]
```diff
@@ -505,6 +508,20 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
             else
                 jl_errorf("julia: invalid argument to --handle-signals (%s)", optarg);
             break;
+        case 's':
+            {
+                int s = strtol(optarg, &endptr, 10);
+                if (errno != 0 || optarg == endptr || *(endptr + 1) != 0) {
```
Comment (vtjnash) [inline src/jloptions.c]: according to the man page, `errno ... is not portable across all platforms`

Comment (yuyichao) [inline src/jloptions.c]: Portability apart, it is wrong to use errno to check for error. It should only be used to check which error happens if the return value indicates that one does happen.

File: src/gc-pages.c (hunk 9 above)
Comment (nkottary) [inline src/gc-pages.c]: It seems memory exception will be thrown if `pg_cnt <= MIN_REGION_PG_COUNT` in the function `jl_gc_alloc_region`. (Need to change `>=` to `>` in the above line)

File: src/jloptions.c (hunk 4 above)
Comment (nkottary) [inline src/jloptions.c]: Could you briefly describe how I may do this?

Comment (nkottary) [inline src/jloptions.c]: Yes.

**Review by yuyichao — APPROVED** [review]
LGTM from GC/correctness POV.

**Comment by yuyichao** [comment]
(Still need tests though)

**Comment by tkelman** [comment]
and should be reflected on the julia JLOptions side, and documented in the manual and man page

**Comment by amitmurthy** [comment]
`config-gc-region-size` can be shortened to `gc-region-size`.

File: src/jloptions.c [hunk 11]
```diff
@@ -181,6 +186,8 @@ JL_DLLEXPORT void jl_parse_opts(int *argcp, char ***argvp)
         { "polly",           required_argument, 0, opt_polly },
         { "math-mode",       required_argument, 0, opt_math_mode },
         { "handle-signals",  required_argument, 0, opt_handle_signals },
+        // config options
+        { "config.gc.region.size", required_argument, 0, opt_config_gc_region_size },
```
Comment (vtjnash) [inline src/jloptions.c]: They aren't really a phrase; I suggested `.` because they are intended as a hierarchal grouping separator.

**Comment by vtjnash** [comment]
I think it should be either `--config.gc.region-size` or `--config.gc.region.size` (depending on whether we expect there to be other configuration properties for `region`)

File: test/cmdlineargs.jl [hunk 12]
```diff
@@ -28,6 +28,9 @@ let exename = `$(Base.julia_cmd()) --precompiled=yes --startup-file=no`
     @test success(`$exename -H $JULIA_HOME`)
     @test success(`$exename --home=$JULIA_HOME`)
 
+    # --config-gc-region-size
+    @test success(`$exename --config-gc-region-size=256M`)
```
Comment (tkelman) [inline test/cmdlineargs.jl]: should read the value inside the julia process to verify that it was set correctly

File: doc/man/julia.1 [hunk 13]
```diff
@@ -84,6 +84,10 @@ Load ~/.juliarc.jl
 Enable or disable Julia's default signal handlers
 
 .TP
+--config-gc-region-size={N}
+Set the garbage collector's region size
```
Comment (tkelman) [inline doc/man/julia.1]: describe briefly the units of the value

**Review by tkelman — DISMISSED** [review]

File: doc/man/julia.1 [hunk 14]
```diff
@@ -85,7 +85,7 @@ Enable or disable Julia's default signal handlers
 
 .TP
 --config-gc-region-size={N}
-Set the garbage collector's region size
+Set the garbage collector's region size in megabytes (suffix with 'M') or gigabytes (suffix with 'G')
```
Comment (tkelman) [inline doc/man/julia.1]: everywhere it's documented, not just here

File: doc/manual/getting-started.rst [hunk 15]
```diff
@@ -122,6 +122,8 @@ those available for the ``perl`` and ``ruby`` programs:
      --startup-file={yes|no}   Load ~/.juliarc.jl
      --handle-signals={yes|no} Enable or disable Julia's default signal handlers
 
+     --config-gc-region-size={N}  Set the garbage collector's region size
```
Comment (tkelman) [inline doc/manual/getting-started.rst]: describe units consistently in all of the documentation

**Comment by nkottary** [comment]
The AppVeyor failure does not look like it is caused by the changes in this PR.

File: src/jloptions.c (hunk 4 above)
Comment (nkottary) [inline src/jloptions.c]: Should we do this in a different PR?

**Comment by vtjnash** [comment]
`region_pg_cnt` was removed years ago
